### PR TITLE
Issue 0091

### PR DESCRIPTION
**Pull request by Nyan11**
https://github.com/OpenSmock/Pyramid/issues/91

### CODE REVIEW

**Comment by Nyan11** [comment]
The CI is fix here : https://github.com/OpenSmock/Bloc-Serialization/pull/11

**Review by labordep — COMMENTED** [review]
I don't understand why "roots" have an "s" and why there is a collection and not just BlSpace>>root element when there is no selection.

image

**Comment by Nyan11** [comment]
> I don't understand why "roots" have an "s" and why there is a collection and not just BlSpace>>root element when there is no selection.
> 

Hello,

I choose "root" to be the collection containing the BlElement.

It was a requirement that Pyramid saves a Collection of elements.
 
If we make it able to edit the `BlSpace>>root element`, i'm afraid we could have unwanted behaviour with the openFromSpace shortcut.

Is the `BlSpace>>root element` the same in Pyramid and in the Space we opened it from ?

**Comment by labordep** [comment]
> > I don't understand why "roots" have an "s" and why there is a collection and not just BlSpace>>root element when there is no selection.
> 
> Hello,
> 
> I choose "root" to be the collection containing the BlElement.
> 
> It was a requirement that Pyramid saves a Collection of elements.
> 
> If we make it able to edit the `BlSpace>>root element`, i'm afraid we could have unwanted behaviour with the openFromSpace shortcut.
> 
> Is the `BlSpace>>root element` the same in Pyramid and in the Space we opened it from ? (We have 2 BlSpace and also 2 root BlElement. One from the original space, one to display the Pyramid Editor)

Hi @Nyan11!

We shouldn't show users the technical details of Pyramid. 
So, technically, it's good to handle two "Spaces": one for editing and one for viewing. 
We can improve this later if needed.

Why does it matter? 
From a user perspective, in the Playground, when I haven't selected elements, I use a Space and add elements to it. 
As a user, I don't know the technical details of Pyramid, and I don't want to. 
Pyramid users expect a Bloc API level in the Playground. 
When no elements are selected, I can access the "root" element of the BlSpace (without an "s" because this is a Pyramid technical stuff).

If I understand your explaination there is an issue because if we expose the root to the user, this is not the "view" root but the pyramid edition "root", right ? The risk is to install behavior on this root and lost it when the view is opened in a test window or when the editor is closed to go back the original edited view ?

**Comment by Nyan11** [comment]
> If I understand your explaination there is an issue because if we expose the root to the user, this is not the "view" root but the pyramid edition "root", right ? The risk is to install behavior on this root and lost it when the view is opened in a test window or when the editor is closed to go back the original edited view ?

Yes. For example, i set a specific layout in my root to display the children, this layout will not be replicated in "test space" or "original space".

**Comment by Nyan11** [comment]
One possible solution is to force the user to always have always one element as root.

The +++
* the root can be the same as the "original" root and the "test" root
* It make serializing/materializing a bit simplier.

The ---
* You cannot have multiples elements as roots
* It will change a lot of thing in Pyramid
* If you want to replace the root element, it will be complicated. (For example if you want your root to be a (ToPanel and not a BlElement)

File: src/Pyramid/PyramidSpCodeObjectInteractionModel.class.st [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+	self projectModel selection size = 1 ifTrue: [
+		^ self projectModel selection first ].
+	^ self projectModel selection
+]
+
+{ #category : #accessing }
+PyramidSpCodeObjectInteractionModel >> doItReceiverString [
+
+	self projectModel selection ifEmpty: [
+		^ 'roots (use #Collection methods)' ].
```
Comment (labordep) [inline src/Pyramid/PyramidSpCodeObjectInteractionModel.class.st:23]: Change term "roots" to "first level elements"

**Review by labordep — CHANGES_REQUESTED** [review]

File: src/Pyramid/PyramidSpCodeObjectInteractionModel.class.st [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+	self projectModel selection size = 1 ifTrue: [ ^ 'a single element' ].
```
Comment (labordep) [inline src/Pyramid/PyramidSpCodeObjectInteractionModel.class.st:24]: "a single" -> "the selected"

File: src/Pyramid/PyramidSpCodeObjectInteractionModel.class.st [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+	^ 'a collection of <1p> elements' expandMacrosWith:
```
Comment (labordep) [inline src/Pyramid/PyramidSpCodeObjectInteractionModel.class.st:25]: "a collection of" -> "an Array of selected "

**Review by labordep — CHANGES_REQUESTED** [review]

**Comment by Nyan11** [comment]
Chec this work : https://github.com/OpenSmock/Pyramid/issues/111

**Comment by Nyan11** [comment]
image

image

image

image

image

**Review by labordep — APPROVED** [review]

**Pull request merged**